### ISSUE TITLE
(acm): Ability to generate and import self-signed certificates to ACM

### ISSUE THREAD

**Issue by tmokmss** [aws/aws-cdk#21115]
### Describe the feature

Like [the feature implemented on Terraform](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate#existing-certificate-body-import), we'd like to generate a self-signed certificate within CDK, and import it to ACM.

The API on Terraform is like below. We can basically follow this design.

```tf
resource "tls_private_key" "example" {
  algorithm = "RSA"
}

resource "tls_self_signed_cert" "example" {
  key_algorithm   = "RSA"
  private_key_pem = tls_private_key.example.private_key_pem

  subject {
    common_name  = "example.com"
    organization = "ACME Examples, Inc"
  }

  validity_period_hours = 12

  allowed_uses = [
    "key_encipherment",
    "digital_signature",
    "server_auth",
  ]
}

resource "aws_acm_certificate" "cert" {
  private_key      = tls_private_key.example.private_key_pem
  certificate_body = tls_self_signed_cert.example.cert_pem
}
```

### Use Case

Currently we cannot create an ACM certificate without 1. creating a Route 53 hosted zone or 2. creating a private certificate authority ([doc](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_certificatemanager-readme.html#requesting-private-certificates)). When I want to quickly test some TLS feature, I sometimes get frustrated since creating them properly is troublesome and difficult.

Although we can generate a self-signed certificate locally and import it via the management console or AWS API, it would be really great if we could do it just using CDK.

### Proposed Solution

Because CloudFormation does not support natively to import a certificate to ACM, we have to create a custom resource construct to do that. Additionally, we need a feature to generate a TLS certificate [as in Terrafrom](https://registry.terraform.io/providers/hashicorp/tls/latest/docs/resources/self_signed_cert) and pass it to the custom resource.

### Other Information

Since using self-signed certificates is often not following security best practice, a warning should be clearly noted like in Terraform docs.

> Self-signed certificates are usually used only in development environments or applications deployed internally to an organization. Certificates of this type are generally not trusted by client software such as web browsers. Therefore clients are likely to generate trust warnings when connecting to a server that has a self-signed certificate.

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.26.0

### Environment details (OS name and version, etc.)

macOS

**Comment by peterwoodworth** [comment]
Thanks for the suggestion!

I am marking this issue as p2, which means that we are unable to work on this immediately.

We use +1s to help prioritize our work, and are happy to revaluate this issue based on community feedback. You can reach out to the cdk.dev community on [Slack](https://join.slack.com/t/cdk-dev/shared_invite/zt-mso6p56d-qJp7SOTBvMaQuDrx7R2wHg) to solicit support for reprioritization.

Check out our [contributing guide](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md) if you're interested in contributing yourself - there's a low chance the team will be able to address this soon but we'd be happy to review a PR 🙂

**Comment by ADiTuri** [comment]
Interested in the feature as well, any news if this is in the roadmap?

**Comment by pahud** [comment]
Hi @tmokmss,

Thank you for this well-detailed feature request! The ability to work with self-signed certificates in ACM is indeed valuable for development and testing scenarios, and the significant community interest (43 upvotes!) demonstrates this is a real need.

After analyzing the current [Certificate](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-certificatemanager/lib/certificate.ts#L280) implementation and considering the proposed solution, we'd like to share our perspective on the best approach to address this use case.

## Our Recommendation: Certificate Import, Not Generation

While your proposed solution of using custom resources for certificate generation is technically feasible, we believe there's a better approach that follows security best practices and reduces maintenance complexity:

### Why We Don't Recommend CDK-Based Certificate Generation

1. **Security Best Practices**: Private keys should never exist in cloud runtime environments like Lambda. Industry standards recommend generating cryptographic material locally where it can be properly secured and controlled.

2. **Separation of Concerns**: Certificate generation is fundamentally a cryptographic operation (OpenSSL's domain), while infrastructure deployment is CDK's domain. Mixing these creates unnecessary coupling.

3. **Maintenance Burden**: Custom resources for certificate generation would require ongoing maintenance for cryptographic library updates, security patches, and Lambda runtime compatibility.

4. **Performance**: No Lambda cold starts, custom resource timeouts, or deployment delays.

### What We Propose Instead

**Step 1: Generate Certificate Locally (30 seconds)**
```bash
# Standard OpenSSL command
openssl req -x509 -newkey rsa:2048 -keyout private-key.pem -out certificate.pem \
 -days 365 -nodes -subj "/CN=example.com" \
 -addext "subjectAltName=DNS:example.com,DNS:*.example.com"
```
**Step 2: Import via CDK (Clean API)**
```typescript
// What we should add to CDK
const certificate = new acm.ImportedCertificate(this, 'SelfSignedCert', {
 certificateBody: fs.readFileSync('./certs/certificate.pem', 'utf8'),
 privateKey: fs.readFileSync('./private-key.pem', 'utf8'),
});

// Works seamlessly with all AWS services
new elbv2.ApplicationLoadBalancer(this, 'ALB', {
 // ... other props
 certificates: [certificate]
});
```
### Benefits of This Approach

- ✅ **Security**: Private keys never leave your local environment
- ✅ **Simplicity**: No custom Lambda code for CDK team to maintain
- ✅ **Flexibility**: Users can use any certificate generation tool (OpenSSL, node-forge, custom scripts)
- ✅ **Performance**: No runtime overhead or custom resource complexity
- ✅ **Standards**: Follows industry cryptographic best practices

### Implementation Plan

We propose implementing an `ImportedCertificate` construct that:
1. Provides a clean L2 interface around CloudFormation's certificate import capability
2. Implements the `ICertificate` interface for seamless integration
3. Includes comprehensive documentation with certificate generation examples
4. Adds appropriate security warnings about self-signed certificate limitations

This gives you the functionality you need (easy ACM import for development) while maintaining security best practices and minimizing long-term maintenance burden.


### `ImportedCertificate` construct

We do not already have an `acm.ImportedCertificate` construct. What we currently have is:

1. `Certificate.fromCertificateArn()` - A static method that imports an existing certificate by ARN
2. `Certificate` class - For creating new certificates with DNS/email validation
3. `PrivateCertificate` class - For certificates from private certificate authorities

The `Certificate.fromCertificateArn()` method creates an imported certificate instance, but it's not a separate ImportedCertificate class. As shown in the code:

```typescript
public static fromCertificateArn(scope: Construct, id: string, certificateArn: string): ICertificate {
  class Import extends CertificateBase {
    public readonly certificateArn = certificateArn;
  }
  return new Import(scope, id);
}
```

This method creates an anonymous class that extends `CertificateBase` and implements the `ICertificate` interface.

The proposed `ImportedCertificate` construct would be a new addition to support importing certificates that were created outside of CDK (including self-signed certificates generated manually). This would complement the existing
`Certificate.fromCertificateArn()` method by providing a more explicit construct for imported certificates, potentially with additional properties and methods specific to imported certificates.

The current `fromCertificateArn()` method works well for referencing existing certificates, but a dedicated `ImportedCertificate` construct could provide better developer experience and clearer intent when working with externally-generated certificates.

## Next Steps

This would be a much smaller implementation effort than the custom resource approach, while still solving the core use case. 

We believe this approach better serves the community's needs while following security and architectural best practices. What are your thoughts on this alternative approach?

Thanks again for bringing this important use case to our attention!

**Comment by tmokmss** [comment]
Thanks @pahud for the detailed suggestion! That sounds good to me, and I agree with the difficulty to support and maintain a new certificate library by the cdk team.

As for generating a certificate part, I think the community can publish a custom resource construct library dedicated for the purpose. Then we can pass the output to the acm.ImportedCertificate construct.

**Comment by pahud** [comment]
@tmokmss  Yes I am working on the PR for `ImportedCertificate` construct now. Will publish the draft PR very soon.
